Finish off span adoption in RemoteGraphicsContextGL.h

geoffreygaren opened the pull request:
#### 254d15e9ef1e495f02f8f1a9e8dec6960fc4852d
<pre>
Finish off span adoption in RemoteGraphicsContextGL.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=282709">https://bugs.webkit.org/show_bug.cgi?id=282709</a>
<a href="https://rdar.apple.com/139381582">rdar://139381582</a>

Reviewed by Chris Dumez.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h: No need for
markers anymore.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h: Autogenerated
(getFloatv):
(getIntegerv):
(getIntegeri_v):
(getBooleanv):
(getShaderPrecisionFormat):
(getUniformfv):
(getUniformiv):
(getUniformuiv):
(getActiveUniformBlockiv):
(getInternalformativ):

* Tools/Scripts/generate-gpup-webgl: Use safe span constructors.
(cpp_type_container.is_dynamic_span):
(cpp_type_container):
(cpp_type_container.is_array):
(cpp_array_reinterpret_cast_conversion):

* Source/WTF/wtf/StdLibExtras.h: Reversed the order of U and Extent because it
is useful to be able to specify Extent and infer U.

Canonical link: <a href="https://commits.webkit.org/286282@main">https://commits.webkit.org/286282@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39bfd6d7f9b2588c3c10a26796aa68e9e2fb8e05

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75441 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79918 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26705 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77557 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2672 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/59213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/17418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78508 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64822 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/39570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22311 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25033 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68591 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22647 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81399 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74703 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2780 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1756 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67454 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2931 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64801 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66744 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10692 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8846 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96971 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11650 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2737 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5558 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21189 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2762 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3697 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2769 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
```diff
@@ -379,21 +379,21 @@
         std::array<GCGLint, 4> value { };
         protectedContext()->getIntegeri_v(pname, index, value);
-        completionHandler(std::span<const int32_t, 4>(reinterpret_cast<const int32_t*>(value.data()), value.size()));
+        completionHandler(spanReinterpretCast<const int32_t>(std::span(value)));
```
I think we could have also used `std::span<const GCGLink>(value)` here and avoided the changes to spanReinterpretCast() right?

geoffreygaren replied:
I'm not super familiar with this code; I was just trying to mimic the existing design of the autogenerator.

GCGLint is typedef'd to int, so you can avoid the cast on any platform where int == int32_t, which is probably most platforms. The same appears to be true of GCGLfloat and GCGLuint, at least.

Should I just remove all the casts? fear removing them, since I don't know why they were there to begin with. Maybe I'll break a platform I don't know about.

cdumez replied:
I didn't suggest removing any cases. My concern is that you're constructing a non-const span only to cast it to a const span. In order to achieve that, you had to modify `spanReinterpretCast`. However, it seems you could simply construct a const span right away and `spanReinterpretCast` would only cast between const spans:
```
completionHandler(spanReinterpretCast<const int32_t>(std::span<const GCGLink>(value)));
```

geoffreygaren replied:
I see. I think the autogenerator has enough metadata to do that. I can give it a shot.

Is it bad to cast *to* const?

cdumez replied:
> Is it bad to cast to const?

Not really. Just seemed unfortunate to create a non-const span only to cast it to const right away.

geoffreygaren committed:
Finish off span adoption in RemoteGraphicsContextGL.h
https://bugs.webkit.org/show_bug.cgi?id=282709
rdar://139381582

Reviewed by Chris Dumez.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h: No need for
markers anymore.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h: Autogenerated
(getFloatv):
(getIntegerv):
(getIntegeri_v):
(getBooleanv):
(getShaderPrecisionFormat):
(getUniformfv):
(getUniformiv):
(getUniformuiv):
(getActiveUniformBlockiv):
(getInternalformativ):

* Tools/Scripts/generate-gpup-webgl: Use safe span constructors.
(cpp_type_container.is_dynamic_span):
(cpp_type_container):
(cpp_type_container.is_array):
(cpp_array_reinterpret_cast_conversion):

* Source/WTF/wtf/StdLibExtras.h: Reversed the order of U and Extent because it
is useful to be able to specify Extent and infer U.

Canonical link: https://commits.webkit.org/286282@main
